internal/response_asserter/response_asserter.go

```diff
@@ -72,12 +88,47 @@ func (a ResponseAsserter[ResponseType]) DecodeAndAssertSingleFields(responsePayl
 	return actualResponse, nil
 }
 
-func (a ResponseAsserter[ResponseType]) DecodeAndAssert(responsePayload []byte) (ResponseType, error) {
-	actualResponse, err := a.DecodeAndAssertSingleFields(responsePayload)
+func (a ResponseAsserter[ResponseType]) DecodeAndAssert(response kafka_client.Response) (ResponseType, error) {
+	actualResponse, err := a.DecodeAndAssertSingleFields(response)
 
 	if err != nil {
 		return actualResponse, err
 	}
 
 	return actualResponse, a.Assertion.AssertAcrossFields(actualResponse, a.Logger)
 }
+
+func (a ResponseAsserter[ResponseType]) assertMessageLength(response kafka_client.Response) error {
+	if len(response.RawBytes) < 4 {
+		return fmt.Errorf("Response is not long enough to accomodate message size.")
+	}
+
+	messageSize := int32(binary.BigEndian.Uint32(response.RawBytes[0:4]))
+
+	if messageSize == int32(len(response.Payload)) {
+		return nil
+	}
+
+	a.Logger.Errorln("❌ Invalid response:")
+	a.Logger.Errorln("The Message Size field does not match the length of the received payload.")
+
+	a.Logger.Errorln("")
+
+	a.Logger.Errorln("🔍 Mismatch:")
+	a.Logger.Errorf(
+		"Message Size field:\t\t%d (Bytes: %02x %02x %02x %02x)",
+		messageSize,
+		response.RawBytes[0], response.RawBytes[1], response.RawBytes[2], response.RawBytes[3],
+	)
+	a.Logger.Errorf("Received payload length:\t%d", len(response.Payload))
+
+	possiblyIncludesMessageSize := messageSize == int32(len(response.Payload)+4)
+
+	if possiblyIncludesMessageSize {
+		a.Logger.Errorln("")
+		a.Logger.Errorln("💡 Hint:")
+		a.Logger.Errorln("The Message Size field should not count itself.")
+	}
+
+	return errors.New("")
```

Comment:
Don't think I've seen us do this elsewhere. Does this cause an empty line to be printed out?